### PR TITLE
Prevent admin deletion and edition 

### PR DESCRIPTION
**Pull request by tuka217**

### CODE REVIEW

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        $this->session = $session;
+    }
+
+    /**
+     * @param GenericEvent $event
+     */
+    public function preDelete(GenericEvent $event)
+    {
+        if (!($subject = $event->getSubject()) instanceof AdminUserInterface) {
+            throw new UnexpectedTypeException($subject, 'Sylius\Component\Core\Model\AdminUserInterface');
```
Comment (michalmarcinkowski) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: Use `AdminUserInterface::class`

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        if ('sylius@example.com' === $subject->getEmail()) {
+            $this->session->getFlashBag()->add('error', 'sylius_demo.account.prevent_delete');
```
Comment (michalmarcinkowski) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: How about `sylius_demo.admin_account.prevent_delete`?

File: src/AppBundle/Resources/config/services.yml [hunk 3]
```diff
@@ -1 +1,11 @@
-services: ~
\ No newline at end of file
+services:
+    sylius_demo.listener.admin_delete:
+        class: AppBundle\EventListener\AdminDeleteListener
+        arguments: ['@session', '@translator']
+        tags:
+            - { name: kernel.event_listener, event: sylius.admin_user.pre_delete, method: preDelete }
+    sylius_demo.listener.admin_update:
```
Comment (michalmarcinkowski) [inline src/AppBundle/Resources/config/services.yml]: Blank line before new service

File: src/AppBundle/Resources/config/services.yml [hunk 4]
```diff
@@ -1 +1,11 @@
...
+        class: AppBundle\EventListener\AdminUpdateListener
```
Comment (michalmarcinkowski) [inline src/AppBundle/Resources/config/services.yml]: Actually we can squash this two listeners into a single listener with 2 methods.

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        if (!($subject = $event->getSubject()) instanceof AdminUserInterface) {
+            throw new UnexpectedTypeException($subject, 'Sylius\Component\Core\Model\AdminUserInterface');
+        }
+        
+        if ('sylius@example.com' === $subject->getEmail()) {
+            $this->session->getFlashBag()->add('error', 'sylius_demo.account.prevent_edit');
+            $event->stopPropagation();
+        }
+    }
+}
```
Comment (michalmarcinkowski) [inline src/AppBundle/EventListener/AdminUpdateListener.php]: Blank line!

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+class AdminDeleteListener
+{
+    /**
+     * @var Session
+     */
+    private $session;
+
+    /**
+     * Constructor.
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: :taco:

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\File\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @author Mateusz Zalewski <mateusz.p.zalewski@gmail.com>
+ */
+class AdminDeleteListener
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: final :>

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $this->session = $session;
+    }
+
+    /**
+     * @param GenericEvent $event
+     */
+    public function preDelete(GenericEvent $event)
+    {
+        if (!($subject = $event->getSubject()) instanceof AdminUserInterface) {
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminDeleteListener.php]:
```php
$subject = $subject = $event->getSubject();
if (!$subject instanceof AdminUserInterface) {
  // ...
}
```

File: src/AppBundle/EventListener/AdminDeleteListener.php (hunk 6 above)
Comment (Zales0123) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: Redundant

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            throw new UnexpectedTypeException($subject, 'Sylius\Component\Core\Model\AdminUserInterface');
+        }
+
+        if ('sylius@example.com' === $subject->getEmail()) {
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: `$subject->getEmailCanonical()`

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ */
+class AdminUpdateListener
+{
+    /**
+     * @var Session
+     */
+    private $session;
+   
+    /**
+     * Constructor.
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminUpdateListener.php]: 🌮

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\File\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class AdminUpdateListener
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminUpdateListener.php]: final :>

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $this->session = $session;
+    }
+    
+    /**
+     * @param GenericEvent $event
+     */
+    public function preUpdate(GenericEvent $event)
+    {
+        if (!($subject = $event->getSubject()) instanceof AdminUserInterface) {
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminUpdateListener.php]:
```php
$subject = $subject = $event->getSubject();
if (!$subject instanceof AdminUserInterface) {
  // ...
}
```

File: src/AppBundle/EventListener/AdminDeleteListener.php [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace AppBundle\EventListener;
+
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\File\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @author Mateusz Zalewski <mateusz.p.zalewski@gmail.com>
```
Comment (Zales0123) [inline src/AppBundle/EventListener/AdminDeleteListener.php]: Why do you impersonate me, damsel?

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+            throw new UnexpectedTypeException($subject, 'Sylius\Component\Core\Model\AdminUserInterface');
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminUpdateListener.php]: `AdminUserInterface::class`

File: src/AppBundle/EventListener/AdminUpdateListener.php [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        
+        if ('sylius@example.com' === $subject->getEmail()) {
```
Comment (pamil) [inline src/AppBundle/EventListener/AdminUpdateListener.php]: `$subject->getEmailCanonical()`

File: src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\File\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class AdminDeleteListenerSpec extends ObjectBehavior
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php]: final :>

File: src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('AppBundle\EventListener\AdminDeleteListener');
+    }
+
+    function it_is_triggered_pre_delete(
+        GenericEvent $event,
+        AdminUserInterface $adminUser,
+        FlashBagInterface $flashBag,
+        $session
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php]: `Session $session`, also it should be the first argument

File: src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        $event->stopPropagation()->shouldBeCalled();
+        
+        $this->preDelete($event);
+    }
+
+    function it_throws_exception_if_event_subject_is_not_customer(GenericEvent $event)
+    {
+        $event->getSubject()->willReturn('badObject');
+        
+        $this->shouldThrow(new UnexpectedTypeException('badObject', 'Sylius\Component\Core\Model\AdminUserInterface'))->during('preDelete', array($event));
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php]: `::class`

File: src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('AppBundle\EventListener\AdminUpdateListener');
+    }
+
+    function it_is_triggered_pre_update(
+        GenericEvent $event,
+        AdminUserInterface $adminUser,
+        FlashBagInterface $flashBag,
+        $session
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php]: `Session $session`, also it should be the first argument

File: src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\File\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class AdminUpdateListenerSpec extends ObjectBehavior
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php]: final :>

File: src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+{
+    function let(Session $session)
+    {
+        $this->beConstructedWith($session);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('AppBundle\EventListener\AdminUpdateListener');
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php]: `::class`

File: src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+        $flashBag->add('error', 'sylius_demo.account.prevent_edit')->shouldBeCalled();
+        $event->stopPropagation()->shouldBeCalled();
+
+        $this->preUpdate($event);
+    }
+
+    function it_throws_exception_if_event_subject_is_not_customer(GenericEvent $event)
+    {
+        $event->getSubject()->willReturn('badObject');
+        $this->shouldThrow(new UnexpectedTypeException('badObject', 'Sylius\Component\Core\Model\AdminUserInterface'))->during('preUpdate', array($event));
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminUpdateListenerSpec.php]: `::class`

File: src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+class AdminDeleteListenerSpec extends ObjectBehavior
+{
+    function let(Session $session)
+    {
+        $this->beConstructedWith($session);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('AppBundle\EventListener\AdminDeleteListener');
```
Comment (pamil) [inline src/AppBundle/spec/EventListener/AdminDeleteListenerSpec.php]: `::class`

**Comment by pjedrzejewski** [comment]
Thank you Ania!

**Pull request merged**